File: slack-reader/src/test/java/dev/number6/slackreader/SnsNotificationAdaptorTest.java
package dev.number6.slackreader;

import dev.number6.slackreader.adaptor.EnvironmentVariableSlackReaderConfigurationAdapter;
import dev.number6.slackreader.adaptor.SnsNotificationAdaptor;
import dev.number6.slackreader.port.SlackReaderConfigurationPort;
import dev.number6.slackreader.port.NotificationPort;
import dev.number6.slackreader.generate.SlackReaderRDG;
import dev.number6.message.ChannelMessages;
import com.amazonaws.services.sns.AmazonSNS;
import com.amazonaws.services.sns.model.PublishRequest;
import com.google.gson.Gson;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import java.time.LocalDate;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

class SnsNotificationAdaptorTest {

    private static final String TOPIC_ARN = "Topic-Arn";

    private final Gson gson = new Gson();
    private final SlackReaderConfigurationPort config = mock(EnvironmentVariableSlackReaderConfigurationAdapter.class);
    private final AmazonSNS sns = mock(AmazonSNS.class);
    private NotificationPort testee;

    @BeforeEach
    void setup() {
        when(config.getTopicArn()).thenReturn(TOPIC_ARN);
        testee = new SnsNotificationAdaptor(config, sns);
    }

    @Test
    void sendSerializedChannelMessagesToSns() {
        ChannelMessages messages = new ChannelMessages(SlackReaderRDG.string().next(), SlackReaderRDG.list(SlackReaderRDG.string()).next(), LocalDate.now());
        testee.broadcast(messages);

        ArgumentCaptor<PublishRequest> captor = ArgumentCaptor.forClass(PublishRequest.class);
        verify(sns).publish(captor.capture());
        assertThat(captor.getValue().getTopicArn()).isEqualTo(TOPIC_ARN);
        assertThat(gson.fromJson(captor.getValue().getMessage(), ChannelMessages.class)).isEqualTo(messages);
    }
}

File: slack-reader/src/test/kotlin/dev/number6/slackreader/SnsNotificationAdaptorTest.kt
package dev.number6.slackreader

import assertk.assertThat
import assertk.assertions.isEqualTo
import com.amazonaws.services.sns.AmazonSNS
import com.amazonaws.services.sns.model.PublishRequest
import com.google.gson.Gson
import dev.number6.message.ChannelMessages
import dev.number6.slackreader.adaptor.EnvironmentVariableSlackReaderConfigurationAdapter
import dev.number6.slackreader.adaptor.SnsNotificationAdaptor
import dev.number6.slackreader.port.NotificationPort
import dev.number6.slackreader.port.SlackReaderConfigurationPort
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import uk.org.fyodor.generators.RDG
import java.time.LocalDate

internal class SnsNotificationAdaptorTest {
    private val gson = Gson()
    private val config: SlackReaderConfigurationPort = mockk<EnvironmentVariableSlackReaderConfigurationAdapter>(relaxUnitFun = true)
    private val sns = mockk<AmazonSNS>(relaxed = true)
    private var testee: NotificationPort? = null

    @BeforeEach
    fun setup() {
        every { config.getTopicArn() } returns TOPIC_ARN
        testee = SnsNotificationAdaptor(config, sns)
    }

    @Test
    fun sendSerializedChannelMessagesToSns() {
        val messages = ChannelMessages(RDG.string().next(), RDG.list(RDG.string()).next(), LocalDate.now())
        testee?.broadcast(messages)
        val slot = slot<PublishRequest>()
        verify {
            sns.publish(capture(slot))
        }
        assertThat(slot.captured.topicArn).isEqualTo(TOPIC_ARN)

        assertThat(gson.fromJson(slot.captured.message, ChannelMessages::class.java)).isEqualTo(messages)
    }

    companion object {
        private const val TOPIC_ARN: String = "Topic-Arn"
    }
}